### PR TITLE
[compiler] Debug tool to emit change detection code rather than memoization

### PR DESCRIPTION
**Pull request by mvitousek**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #29658
* #29657
* #29653
* __->__ #29656

Summary: The essential assumption of the compiler is that if the inputs to a computation have not changed, then the output should not change either--computation that the compiler optimizes is idempotent.

This is, of course, known to be false in practice, because this property rests on requirements (the Rules of React) that are loosely enforced at best. When rolling out the compiler to a codebase that might have rules of react violations, how should developers debug any issues that arise?

This diff attempts one approach to that: when the option is set, rather than simply skipping computation when dependencies haven't changed, we will *still perform the computation*, but will then use a runtime function to compare the original value and the resultant value. The runtime function can be customized, but the idea is that it will perform a structural equality check on the values, and if the values aren't structurally equal, we can report an error, including information about what file and what variable was to blame.

This assists in debugging by narrowing down what specific computation is responsible for a difference in behavior between the uncompiled code and the program after compilation.

### CODE REVIEW

File: compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/useState-pruned-dependency-change-detect.expect.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  let t0;
+  if ($[0] === Symbol.for("react.memo_cache_sentinel")) {
+    t0 = f(props.x);
+    $[0] = t0;
+  } else {
+    t0 = $[0];
+  }
+  const [x] = useState(t0);
+  let t1;
+  let old$t1;
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/useState-pruned-dependency-change-detect.expect.md:33]: can we move this into the if where it's used?

**Review by josephsavona — APPROVED** [review]
nice. this is mutually exclusive with the option to always emit `|| true` in the if check, right? let's enforce that if so.

File: compiler/packages/snap/src/compiler.ts [hunk 2]
```diff
@@ -128,6 +129,12 @@ function makePluginOptions(
   if (firstLine.includes("@disableMemoizationForDebugging")) {
     disableMemoizationForDebugging = true;
   }
+  if (firstLine.includes("@enableChangeDetectionForDebugging")) {
+    enableChangeDetectionForDebugging = {
+      source: "react-compiler-runtime",
```
Comment (poteto) [inline compiler/packages/snap/src/compiler.ts:134]: fyi you might already know this but we don't currently publish this package on npm, so this flag will only work internally for now

Comment (mvitousek) [reply compiler/packages/snap/src/compiler.ts:134]: Yeah, I wanted to include the runtime code as basically an example of/documentation for what the change detection code would look like. I've been updating it internally separately

**Review by josephsavona — APPROVED** [review]
nice

**Pull request merged**